ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -73,25 +73,62 @@ private class ContractsFetch(
   )(implicit
       ec: ExecutionContext,
       mat: Materializer,
-  ): ConnectionIO[List[BeginBookmark[domain.Offset]]] = {
-    import cats.instances.list._, cats.syntax.traverse._, doobie.implicits._, doobie.free.{
-      connection => fc
-    }
+  ): ConnectionIO[BeginBookmark[Terminates.AtAbsolute]] = {
+    import cats.instances.list._, cats.syntax.foldable.{toFoldableOps => ToFoldableOps},
+    cats.syntax.traverse.{toTraverseOps => ToTraverseOps}, cats.syntax.functor._, doobie.implicits._
     // we can fetch for all templateIds on a single acsFollowingAndBoundary
     // by comparing begin offsets; however this is trickier so we don't do it
     // right now -- Stephen / Leo
     connectionIOFuture(getTermination(jwt)) flatMap {
-      _ cata (absEnd =>
-        templateIds.traverse {
-          fetchAndPersist(jwt, parties, absEnd, _)
-        }, fc.pure(templateIds map (_ => LedgerBegin)))
+      _.cata(
+        absEnd =>
+          // traverse once, use the max _returned_ bookmark,
+          // re-traverse any that != the max returned bookmark (overriding lastOffset)
+          // fetch cannot go "too far" the second time
+          templateIds.traverse(fetchAndPersist(jwt, parties, None, absEnd, _)).flatMap {
+            actualAbsEnds =>
+              val newAbsEndTarget = {
+                import scalaz.std.list._, scalaz.syntax.foldable._,
+                domain.Offset.{ordering => `Offset ordering`}
+                // it's fine if all yielded LedgerBegin, so we don't want to conflate the "fallback"
+                // with genuine results
+                actualAbsEnds.maximum getOrElse AbsoluteBookmark(absEnd.toDomain)
```

Comment:
We only reach the fallback if the list of template ids is empty right?

Reply:
Right.